src/components/OTPInput/OTPInput.tsx

```diff
@@ -0,0 +1,98 @@
+// lib
+import { FormEvent, FormEventHandler, useRef, useState } from "react";
+
+// components
+import DisplayInput from "../DisplayInput";
+
+// styles
+import styles from "./styles.module.css";
+
+// helpers
+import { isNumber } from "../../utils";
+
+// hooks
+import useOutsideClick from "../../hooks/useOutsideClick";
+
+function OTPInput({
+  id = "",
+  label = "",
+  readonly = false,
+  value = "",
+  callback = () => {},
+  error = "",
+}: {
+  id: string;
+  label?: string;
+  readonly?: boolean;
+  value?: string;
+  callback?: (value: string) => void;
+  error?: string | null;
+}) {
+  const inputRef = useRef(null);
+  const [input, setInput] = useState<string>(value);
```

Comment:
My rationale for this approach:
- The OTPInput component is made as a controlled input. It should encapsulate all the display state & error/success states within the component itself. This component can be consumed at a repo level. This component was built keeping generic usecases in mind rather than just to solve the assignment.

Probably not required for this component as OTP inputs don't have much reusability. Such approach can work for normal Inputs which have various usecases.

- In our usecase, we had to make the validation of OTP at parent level. Thus I had to add a callback function to get the state from child. That state is then verified with given OTP and error states are updated. 

I do agree with you. It can be approached in a simpler way. By just keeping the all state at parent and pass the props to child.

- The error state is also duplicated in this case as validation is happening at parent, rather than the component is itself. This can be solved by passing a validation function to the controlled input. Validation function can be then run at controlled input's onchange.